### ISSUE TITLE
"Sylvester's sequence" RC example not working on Windows

### ISSUE THREAD

**Issue by drkameleon** [arturo-lang/arturo#979]
...

**Comment by RickBarretto** [comment]
Strange...

Running on `@full 0.9.83 b/4`
![image](https://user-images.githubusercontent.com/78623871/218887531-fc36c887-49ef-4575-a757-2199f8aa2f7e.png)
Running with `@full 0.9.80`
![image](https://user-images.githubusercontent.com/78623871/218887803-63cd4cfa-8efd-4f03-8f65-c5263d60fae1.png)

**Comment by RickBarretto** [comment]
Testing #978 and #977 I got the same. But it's not even finishing the calculation, so I interrupted the run:
![image](https://user-images.githubusercontent.com/78623871/218889234-b40a7672-f1f6-4597-b38d-3fc51d1e1681.png)
Is there any possibility of relation with #987? Let's see!

**Comment by drkameleon** [comment]
@RickBarretto I totally missed your message.

I'll have a look into it!

**Comment by drkameleon** [comment]
> Strange...
> 
> Running on `@full 0.9.83 b/4` ![image](https://user-images.githubusercontent.com/78623871/218887531-fc36c887-49ef-4575-a757-2199f8aa2f7e.png) Running with `@full 0.9.80` ![image](https://user-images.githubusercontent.com/78623871/218887803-63cd4cfa-8efd-4f03-8f65-c5263d60fae1.png)

Apparently, I did something to break it lol. 

Seriously now, by the look of it, the thing I notice is the 3 last number of the series being wrong (in 0.9.83). And the special thing about them is that they are actually GMP bignums... so it *has* to do with that. But I cannot remember if I changed anything regarding BigNim handling (I would swear I didn't...) Hmm... Let's see...

**Comment by drkameleon** [comment]
Ultra-surprise of the century:

I just tried running Sylvester's sequence with the latest binary (and the correct, updated DLLs), on Windows 10 and...

<img width="939" alt="Screenshot 2023-02-18 at 12 38 47" src="https://user-images.githubusercontent.com/1265028/219860005-04c3ff2f-f9fa-4cc7-8500-28f2d72afd1b.png">

It just works 😲 

Can you please confirm the results with the latest binary? (preferrably, *not* via MSYS; or better put, making sure that the DLLs being used, GMP, MPFR are exactly the ones Arturo comes with)

**Comment by RickBarretto** [comment]
Alright! I can! Update you soon!

**Comment by RickBarretto** [comment]
Running on cmd, from the build you sent me on Discord.
![image](https://user-images.githubusercontent.com/78623871/219883655-793875a0-7156-43ba-8275-363dc396841c.png)

And, running on Msys2:
![image](https://user-images.githubusercontent.com/78623871/219883727-d031abcb-c146-4a90-b80c-f8a8210e66d6.png)

**Comment by RickBarretto** [comment]
@drkameleon, shouldn't this issue be closed? Since it's working now.

**Comment by RickBarretto** [comment]
@drkameleon about this issue. Is it considered solved? Since it is running well for the build you sent me on Discord.

Note:I tested with my latest build, and didn't work. It just works on that specific build you sent me.

**Comment by drkameleon** [comment]
> @drkameleon about this issue. Is it considered solved? Since it is running well for the build you sent me on Discord.
> 
> Note:I tested with my latest build, and didn't work. It just works on that specific build you sent me.

Good point! Ι'll give it another try.

Basically, macOS aside (where I can actually see the errors as they pop up - literally lol), for every other OS my main basis are the Github Actions: if it passes, it's ok.

So, I'll have to try another PR, with the RC example active (basically, rename the `_res` file to `.res`)

**Comment by RickBarretto** [comment]
But, would you change the CI before, wouldn't?
I mean, I don't know exactly what you did to fix it. If was some flag or dll changes...

**Comment by drkameleon** [comment]
Let's see... (I *think* but I'm not sure, it had to do with the binary - Arturo's binary - picking up the "right" DLL; and to make sure, I did in a totally vanilla Windows environment, with nothing installed whatsoever, not even MSYS2...)

I'll first try this in another vanilla machine I have set up right now and will let you know asap. 😉

**Comment by drkameleon** [comment]
So... step-by-step:

- Totally fresh Windows 10 (virtual) machine in the cloud set up (this is taking some time... 😞 )
- I'm using the full binary from here: https://github.com/arturo-lang/arturo/actions/runs/4251293054
- And the code for the specific RC example: https://github.com/arturo-lang/arturo/blob/master/examples/rosetta/sylvester's%20sequence.art (❗ it just struck me that Arturo could even work with scripts as URL, from the command-line: the same way you do `arturo somescript.art`, you should be able to do something like `arturo https://somedomain/somescript.art` - exactly like `do` works for example... Hmmmm....)
- running the script via CMD:
<img width="964" alt="Screenshot 2023-02-23 at 11 54 47" src="https://user-images.githubusercontent.com/1265028/220887175-7db2289f-eb72-4a5e-8007-f288f484d1c7.png">

Result: A disaster lol.

Do you remember which exact binary it was that *was* working?

**Comment by drkameleon** [comment]
Buff... I'm confused. I'm stil getting something different on macOS. So, I guess it's still not working...

I re-checked the expected result from here: https://rosettacode.org/wiki/Sylvester%27s_sequence

And it seems like the Windows version is broken as it is.

So, the question remains: Do you remember which exact binary it was that was working?

**Comment by RickBarretto** [comment]
Wait, is it working from #994? Whaaat

**Comment by drkameleon** [comment]
Let me sum it up...

On macOS and Linux, I'm getting:

```
First 10 terms of the Sylvester sequence:
2 3 7 43 1807 3263443 10650056950807 113423713055421844361000443 12864938683278671740537145998360961546653259485195807 165506647324519964198468195444439180017513152706377497841851388766535868639572406808911988131737645185443 

Sum of the reciprocals of the first 10 items:
1.0
```

This is the only accepted result. And, unfortunately, the Windows version is *still* showing something different...

**Comment by RickBarretto** [comment]
Yes, I remember.


> The latest FULL build for Windows: (Discord doesn't let me upload files > 8MB, so there you go...)
https://arturo-lang.s3.amazonaws.com/testing/arturo-full-x86_64-windows.zip
[On chat](https://discord.com/channels/765519132186640445/829324913097048065/1076197110145351681)

**Comment by drkameleon** [comment]
I tried to see what GMP version is being used on each OS and since we can access this information via Arturo's very own `sys` constant, here it is:

<img width="974" alt="Screenshot 2023-02-23 at 12 12 48" src="https://user-images.githubusercontent.com/1265028/220890924-2775f4fa-1143-4e0f-8b9f-9d995faeabc8.png">

And the GMP version is... identical.

**Comment by RickBarretto** [comment]
Hummm, strange this...

**Comment by drkameleon** [comment]
I'm experimenting a bit...

If you look into the code for this specific example, technically this: `'result ++ inc fold result .seed:1 [a b][a * b]` is the same as `'result ++ inc product result` (I have no idea why I decided to do it with a `fold`).

Despite changing this, the two versions keep working like before (the macOS one does work, while the Windows one doesn't).

Which means that it has nothing to do with the multiplication... but, doesn't it?

If I turn this to `result ++ product result` (which is the same, albeit not what the RC problem asks for), now the results are VERY different...

On macOS:

```
First 10 terms of the Sylvester sequence:
2 2 4 16 256 65536 4294967296 18446744073709551616 340282366920938463463374607431768211456 115792089237316195423570985008687907853269984665640564039457584007913129639936 

Sum of the reciprocals of the first 10 items:
1.0
```

On Windows:

```
First 10 terms of the Sylvester sequence:
2 2 4 16 256 65536 4294967296 0 0 0

Sum of the reciprocals of the first 10 items:
1.0
```

And this looks like some serious mess.

Now, what I guess:

`:integer` values in Arturo are... well... integers, but with a catch; actually it hides 2 different types 64-bit integers and GMP bignums. Roughly speaking, if you do `a: 123`, since 123 fits in a 64-bit integer, internally that's what it will be. If you do `a: 12312736128736128736182638126831`, then since it doesn't fit into a 64-bit integer, it will convert it into a GMP bignum, although you - as a user/programmer I mean - won't notice any difference. Now, what happens with operations? Every "overflow" is silently handled and every such case would be converted into a BigNum. So, let's say you multiply two `:integer` values that are internally 64-bit integers, if the result fits in a 64-bit integer, that's what it will be. But if not, it has to be "upgraded" to a GMP bignum.

Long story short: I'm pretty sure this is exactly the part that is failing. Now, why on Windows only? This remains to be seen... 🤔

**Comment by RickBarretto** [comment]
Thank you for clarifying this!
But, yeah, very strange results...

**Comment by drkameleon** [comment]
I'll open a PR to exclusively check this one... and hopefully fix it.

The culprit should be the `*` (and `*=`) operator, somewhere here: https://github.com/arturo-lang/arturo/blob/master/src/vm/values/value.nim#L1154-L1180
